packages/@uppy/core/src/Uppy.test.ts

```diff
@@ -1472,6 +1472,161 @@ describe('src/Core', () => {
       await core.retryAll()
       expect(onUpload).not.toHaveBeenCalled()
     })
+
+    describe('with required metadata', () => {
+      it('should not retry files that have missing required metadata', async () => {
+        const onUpload = vi.fn()
+        const onRetryAll = vi.fn()
+
+        const core = new Core({
+          restrictions: {
+            requiredMetaFields: ['caption'],
+          },
+        })
+        core.on('upload', onUpload)
+        core.on('retry-all', onRetryAll)
+
+        const fileId = core.addFile({
+          source: 'vi',
+          name: 'foo.jpg',
+          type: 'image/jpeg',
+          data: testImage,
+        })
+
+        // Simulate an upload attempt which triggers metadata validation
+        try {
+          await core.upload()
+        } catch (error) {
+          expect(error).toBeInstanceOf(RestrictionError)
+          expect(error.message).toContain('Missing required meta fields')
+        }
+
+        // File should now have missing metadata error after upload attempt
+        const file = core.getFile(fileId)
+        expect(file.missingRequiredMetaFields).toEqual(['caption'])
+        expect(file.error).toContain('Missing required meta fields in foo.jpg')
+
+        // Should not retry files with outstanding metadata issues
+        await core.retryAll()
+        expect(onRetryAll.mock.calls[0][0]).toEqual([])
+        expect(onUpload).toHaveBeenCalledTimes(0)
+      })
+
+      it('should retry files after metadata is corrected', async () => {
+        const onUpload = vi.fn()
+        const onRetryAll = vi.fn()
+
+        const core = new Core({
+          restrictions: {
+            requiredMetaFields: ['caption'],
+          },
+        })
+        core.on('upload', onUpload)
+        core.on('retry-all', onRetryAll)
+
+        const fileId = core.addFile({
+          source: 'vi',
+          name: 'foo.jpg',
+          type: 'image/jpeg',
+          data: testImage,
+        })
+
+
+        try {
+          await core.upload()
+        } catch (error) {
+          expect(error).toBeInstanceOf(RestrictionError)
+          expect(error.message).toContain('Missing required meta fields')
+        }
+
+        // Verify file has missing metadata error after upload attempt
+        let file = core.getFile(fileId)
+        expect(file.missingRequiredMetaFields).toEqual(['caption'])
+        expect(file.error).toContain('Missing required meta fields in foo.jpg')
+
+        // Fix the metadata
+        core.setFileMeta(fileId, { caption: 'Test caption' })
+
+        // Trigger the dashboard:file-edit-complete event to update validation state
+        // @ts-ignore
+        core.emit('dashboard:file-edit-complete', core.getFile(fileId))
```

Comment:
Does this mean required meta fields do not work outside of dashboard? That would be a design mistake I think 🤔

Reply:
I don't mean the untyped part, I mean can you use metaFields without the dashboard plugin or will the whole thing stop working because there is no `dashboard:file-edit-complete` event?

Reply:
No you can't use metaFields without the dashboard plugin.